README
mhash README
===============================================================================

  OVERVIEW

  Description
  -----------

  The mhash library provides an easy way to access strong hashes such as MD5,
  SHA1 and other algorithms.

  Features
  --------

  The mhash library provides the following features:

   o  Free software (GNU Lesser General Public License)
   o  Reentrancy support
   o  Support for known strong algorithms
   o  Support for less-known algorithms such as GOST
   o  Should work on every ANSI C conforming platform

  Disclaimer
  ----------

  Please read the file COPYING before using mhash.
  
  Installation
  ------------

  For installing mhash under Unix please read the document INSTALL.

  New Versions
  ------------

  Information on how to retrieve new versions of mhash can be found on
  <URL:http://mhash.sourceforge.net>.

  Credits
  -------

  Please read the file AUTHORS.


NEWS
mhash ChangeLog
===============================================================================

People

   NM - Nikos Mavroyanopoulos <nmav@hellug.gr>
   SS - Sascha Schumann <sascha@schumann.cx>

Version 0.9.9
- Added patch for sigbus error on unaligned data, when doing rapid copies.
  (Thanks to Wayne - mhash@useunix.net - for that.)

Version 0.9.8.1
- Another round of bugfixes

Version 0.9.8
- As before - more patches and cleanups

Version 0.9.7
- More patches, more cleanups. Better handling of endian issues. Should
  handle platforms that do weird things with booleans better. Eliminated
  more redundant code. Snefru is now endian-aware. Maybe.

Known Issues:
- Sub-optimal on 64-bit platforms, as a lot of the code assumes 32-bits.
- May be some big-endian issues remaining. make check is your friend!

Version 0.9.6
- Bunch of patches and speedups. Credits to be added.

Version 0.9.2
- Added SNEFRU128, SNEFRU256. Reintroduced MD2.
  (jkmhash@itaparica.org)
- Fixes to .spec file and clean-ups of documentation
  (cpedersen[at]c-note.dk)
- Fixed typecasts, memory leaks and compiler warnings
  (imipak@yahoo.com)

Version 0.9.1
- Added RIPEMD128/256/320, SHA224/384/512, Whirlpool
- Really, this time!
  (jkmhash@itaparica.org)

Version 0.8.19
- Added VC++ project files, contributed by James M. Rogers 
  <james@singlestep.com>

Version 0.8.18 (released March 6 2003)
- Several additions and fixes to built win32 dlls.

Version 0.8.17 (released October 4 2002)
- Corrected the HEX key mode. Refuses to convert hex strings longer
  than the given key size.
- Corrected some memory leaks in keygen_s2k. Patch by Bob Mathews 
  <bobmathews@alumni.calpoly.edu>
- Corrected wrong use of sizeof in hmac_test. Reported by Ellis Pritchard
  <ellispritchard@users.sf.net>

Version 0.8.16 (released May 29 2002)
- Code cleanups and optimizations
- Added configure time options to disable certain algorithms

Version 0.8.15 (released May 24 2002)
- Corrected Iterated S2K
- Made all static tables constant (pointed out by Darryl Miles <darryl@netbauds.net>)

Version 0.8.14 (release March 28 2002)
- mhash_deinit() and mhash_hmac_deinit() now accept a null digest parameter
- Corrected memory leaks (patch by Gustavo Niemeyer <niemeyer@conectiva.com>)
- Corrected bug in Gost hash algorithm (pointed out by Mike Gorchak <mike@malva.ua>)

Version 0.8.13 (released November 18 2001)
- Added ADLER32 algorithm (implementation by Manuel Kasper <mk@neon1.net>)

Version 0.8.12 (released October 29 2001)
- Corrected bugs in mhash_get_name() functions
- Added SHA-256 from libnettle.

Version 0.8.11 (released October 21 2001)
- added mhash_save_state_mem() and mhash_restore_state_mem() functions,
  initial patches and idea by Blake Stephen <Stephen.Blake@veritect.com>
- Added mhash_get_name_static()
- Corrected mhash_get_name() and mhash_count()

Version 0.8.10 (released on July 12 2001)
- New mhash_deinit() and mhash_hmac_deinit() functions based on the
  proposal, and patches by William Ahern <wahern@25thandClement.com>

Version 0.8.9 (released on February 04 2001)
- TIGER runs on alpha (NM)
- Documentation updated

Version 0.8.8 (released on January 29 2001)
- MD4 and MD5 now compile on big endian machines (NM)
- TIGER now compiles on big endian machines. (NM)

Version 0.8.6 (released on January 21 2001)
- Corrected Tiger implementation (NM, bug pointed out by Gordon Mohr <gojomo@usa.net>)
- Added MD4 (NM)
- Added TIGER128 and TIGER160 (NM)

Version 0.8.5 (released on January 2001)
- added checks for malloc failures (NM, some patches by Sami Farin <sfarin@ratol.fi>)
- corrected old style checks for mhash failures (NM, reported by Joe Hohertz <jhohertz@wiretap.net>)

Version 0.8.4 (released on December 15 2000)
- added mhash_hmac_end_m() and mhash_end_m() functions which allow to specify the malloc function. (NM)
- added mhash_cp() - Copy a context (NM)
- added mhash_get_mhash_algo() - Returns the name of the algorithm of the context (NM)
- fixes in crc32b and gost (for multithreaded applications) (NM)

Version 0.8.3 (released on October 25 2000)
- Compiles under Win32 using cygwin (with -mno-cygwin) (NM)
- Bugfixes in KEYGEN_ASIS (NM)

Version 0.8.2 (released on May 12 2000)
- Corrected Memory leaks in HMAC implementation (NM, contributed by Marios Hadjieleftheriou <marioh@cs.ucr.edu>)
- Added HAVAL224 (NM)

Version 0.8.1 (released on April 11 2000)
- Bugfixes in hmac implementation. (NM)
- Removed MD2 (which was not free) (NM)

Version 0.8.0 (released on April 5 2000)
- No longer needs posix thread locks to be thread safe (NM)
- Supports unlimited number of MHASH threads (NM)
- Improved the key generation algorithm API (NM)
- Corrected Haval implementation. (NM)
  Now mhash uses Paulo Barreto's implementation
  Added HAVAL256, HAVAL192, HAVAL160, HAVAL128 (3 passes haval)
- License was changed to GNU Lesser GPL

Version 0.7.0 (released on Mar 3rd 2000)
- cleaned up binary interface (SS)
- improved POSIX Threads detection (SS)
- added keygen API (NM)
- added new improved RIPEMD160 implementation (NM)
- removed RIPEMD128 (NM)
- added RPM .spec file by work@scripty.com (SS)

Version 0.6.1 (released on Oct 23th 1999)
- fixed Tiger algorithm on 32-bit architectures (NM)

Version 0.6.0 (released on Oct 11th 1999)
- export only symbols of the public interface (SS)
- MHASH_API_VERSION has been changed to 19991004 (SS)
- API change: All algorithm names have been prefixed with MHASH_
  to avoid namespace pollution (SS)
- added mhash_get_hash_pblock() by NM, reimplemented by SS
- improved data structure handling (NM)
- use "--silent" libtool option (SS)
- added hmac functions (SS, contributed by NM)

Version 0.5.2 (released on Jul 14th 1999)
- fix VPATH build (SS)
- upgrade to libtool 1.3.3 (SS)
- rework directory layout (simplified) (SS)
- change test script to work with Bourne shells (SS)
- fix some big endian issues in RIPEMD160, RIPEMD128 and TIGER (SS)
	
Version 0.5 (relased on May 22nd 1999)
- put under CVS control

sha1.c
/*
   SHA-1 in C
   By Steve Reid <steve@edmweb.com>
   100% Public Domain

   Test Vectors (from FIPS PUB 180-1)
   "abc"
   A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
   "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
   84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
   A million repetitions of "a"
   34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
 */

/*
   #define LITTLE_ENDIAN * This should be #define'd if true. 
 */
/*
   #define SHA1HANDSOFF * Copies data before messing with it. 
 */

#include "mhash_sha1.h"

#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

/*
   blk0() and blk() perform the initial expand. 
 */
/*
   I got the idea of expanding during the round function from SSLeay 
 */
#ifndef WORDS_BIGENDIAN
#define blk0(i) (block->l[i] = (rol(block->l[i],24)&0xFF00FF00) \
    |(rol(block->l[i],8)&0x00FF00FF))
#else
#define blk0(i) block->l[i]
#endif
#define blk(i) (block->l[i&15] = rol(block->l[(i+13)&15]^block->l[(i+8)&15] \
    ^block->l[(i+2)&15]^block->l[i&15],1))

/*
   (R0+R1), R2, R3, R4 are the different operations used in SHA1 
 */
#define R0(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk0(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R1(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R2(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0x6ED9EBA1+rol(v,5);w=rol(w,30);
#define R3(v,w,x,y,z,i) z+=(((w|x)&y)|(w&x))+blk(i)+0x8F1BBCDC+rol(v,5);w=rol(w,30);
#define R4(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0xCA62C1D6+rol(v,5);w=rol(w,30);


/*
   Hash a single 512-bit block. This is the core of the algorithm. 
 */

void
SHA1Transform(word32 state[5], word8 buffer[64])
{
	word32 a, b, c, d, e;
	typedef union {
		word8 c[64];
		word32 l[16];
	} CHAR64LONG16;
	CHAR64LONG16 *block;
#ifdef SHA1HANDSOFF
	static word8 workspace[64];
	block = (CHAR64LONG16 *) workspace;
	memcpy(block, buffer, 64);
#else
	block = (CHAR64LONG16 *) buffer;
#endif
	/*
	   Copy context->state[] to working vars 
	 */
	a = state[0];
	b = state[1];
	c = state[2];
	d = state[3];
	e = state[4];
	/*
	   4 rounds of 20 operations each. Loop unrolled. 
	 */
	R0(a, b, c, d, e, 0);
	R0(e, a, b, c, d, 1);
	R0(d, e, a, b, c, 2);
	R0(c, d, e, a, b, 3);
	R0(b, c, d, e, a, 4);
	R0(a, b, c, d, e, 5);
	R0(e, a, b, c, d, 6);
	R0(d, e, a, b, c, 7);
	R0(c, d, e, a, b, 8);
	R0(b, c, d, e, a, 9);
	R0(a, b, c, d, e, 10);
	R0(e, a, b, c, d, 11);
	R0(d, e, a, b, c, 12);
	R0(c, d, e, a, b, 13);
	R0(b, c, d, e, a, 14);
	R0(a, b, c, d, e, 15);
	R1(e, a, b, c, d, 16);
	R1(d, e, a, b, c, 17);
	R1(c, d, e, a, b, 18);
	R1(b, c, d, e, a, 19);
	R2(a, b, c, d, e, 20);
	R2(e, a, b, c, d, 21);
	R2(d, e, a, b, c, 22);
	R2(c, d, e, a, b, 23);
	R2(b, c, d, e, a, 24);
	R2(a, b, c, d, e, 25);
	R2(e, a, b, c, d, 26);
	R2(d, e, a, b, c, 27);
	R2(c, d, e, a, b, 28);
	R2(b, c, d, e, a, 29);
	R2(a, b, c, d, e, 30);
	R2(e, a, b, c, d, 31);
	R2(d, e, a, b, c, 32);
	R2(c, d, e, a, b, 33);
	R2(b, c, d, e, a, 34);
	R2(a, b, c, d, e, 35);
	R2(e, a, b, c, d, 36);
	R2(d, e, a, b, c, 37);
	R2(c, d, e, a, b, 38);
	R2(b, c, d, e, a, 39);
	R3(a, b, c, d, e, 40);
	R3(e, a, b, c, d, 41);
	R3(d, e, a, b, c, 42);
	R3(c, d, e, a, b, 43);
	R3(b, c, d, e, a, 44);
	R3(a, b, c, d, e, 45);
	R3(e, a, b, c, d, 46);
	R3(d, e, a, b, c, 47);
	R3(c, d, e, a, b, 48);
	R3(b, c, d, e, a, 49);
	R3(a, b, c, d, e, 50);
	R3(e, a, b, c, d, 51);
	R3(d, e, a, b, c, 52);
	R3(c, d, e, a, b, 53);
	R3(b, c, d, e, a, 54);
	R3(a, b, c, d, e, 55);
	R3(e, a, b, c, d, 56);
	R3(d, e, a, b, c, 57);
	R3(c, d, e, a, b, 58);
	R3(b, c, d, e, a, 59);
	R4(a, b, c, d, e, 60);
	R4(e, a, b, c, d, 61);
	R4(d, e, a, b, c, 62);
	R4(c, d, e, a, b, 63);
	R4(b, c, d, e, a, 64);
	R4(a, b, c, d, e, 65);
	R4(e, a, b, c, d, 66);
	R4(d, e, a, b, c, 67);
	R4(c, d, e, a, b, 68);
	R4(b, c, d, e, a, 69);
	R4(a, b, c, d, e, 70);
	R4(e, a, b, c, d, 71);
	R4(d, e, a, b, c, 72);
	R4(c, d, e, a, b, 73);
	R4(b, c, d, e, a, 74);
	R4(a, b, c, d, e, 75);
	R4(e, a, b, c, d, 76);
	R4(d, e, a, b, c, 77);
	R4(c, d, e, a, b, 78);
	R4(b, c, d, e, a, 79);
	/*
	   Add the working vars back into context.state[] 
	 */
	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;
	state[4] += e;
	/*
	   Wipe variables 
	 */
	a = b = c = d = e = 0;
}


/*
   SHA1Init - Initialize new context 
 */

void
SHA1Init(SHA1_CTX * context)
{
	/*
	   SHA1 initialization constants 
	 */
	context->state[0] = 0x67452301;
	context->state[1] = 0xEFCDAB89;
	context->state[2] = 0x98BADCFE;
	context->state[3] = 0x10325476;
	context->state[4] = 0xC3D2E1F0;
	context->count[0] = context->count[1] = 0;
}


/*
   Run your data through this. 
 */

void
SHA1Update(SHA1_CTX * context, const word8 * data, unsigned int len)
{
	unsigned int i, j;

	j = (context->count[0] >> 3) & 63;
	if ((context->count[0] += len << 3) < (len << 3))
		context->count[1]++;
	context->count[1] += (len >> 29);
	if ((j + len) > 63) {
		memcpy(&context->buffer[j], data, (i = 64 - j));
		SHA1Transform(context->state, context->buffer);
		for (; i + 63 < len; i += 64) {
			SHA1Transform(context->state, (word8 *) & data[i]);
		}
		j = 0;
	}
	else
		i = 0;
	memcpy(&context->buffer[j], &data[i], len - i);
}


/*
   Add padding and return the message digest. 
 */

void *
SHA1Final(SHA1_CTX * context)
{
	word32 i, j;
	word8 finalcount[8];
	word8 *digest = malloc(20);

	for (i = 0; i < 8; i++) {
		finalcount[i] = (word8) ((context->count[(i >= 4 ? 0 : 1)]
								  >> ((3 - (i & 3)) * 8)) & 255);	/*
																	   Endian independent 
																	 */
	}
	SHA1Update(context, (word8 *) "\200", 1);
	while ((context->count[0] & 504) != 448) {
		SHA1Update(context, (word8 *) "\0", 1);
	}
	SHA1Update(context, finalcount, 8);		/*
											   Should cause a SHA1Transform() 
											 */
	for (i = 0; i < 20; i++) {
		digest[i] = (word8)
			((context->state[i >> 2] >> ((3 - (i & 3)) * 8)) & 255);
	}
	/*
	   Wipe variables 
	 */
	i = j = 0;
	memset(context->buffer, 0, 64);
	memset(context->state, 0, 20);
	memset(context->count, 0, 8);
	memset(&finalcount, 0, 8);
#ifdef SHA1HANDSOFF				/* make SHA1Transform overwrite it's own static vars */
	SHA1Transform(context->state, context->buffer);
#endif

	return digest;
}


/*************************************************************/

/*
   int main(int argc, char** argv)
   {
   int i, j;
   SHA1_CTX context;
   word8 *digest, buffer[16384];
   FILE* file;

   if (argc > 2) {
   puts("Public domain SHA-1 implementation - by Steve Reid <steve@edmweb.com>");
   puts("Produces the SHA-1 hash of a file, or stdin if no file is specified.");
   exit(0);
   }
   if (argc < 2) {
   file = stdin;
   }
   else {
   if (!(file = fopen(argv[1], "rb"))) {
   fputs("Unable to open file.", stderr);
   exit(-1);
   }
   } 
   SHA1Init(&context);
   while (!feof(file)) {  
   i = fread(buffer, 1, 16384, file);
   SHA1Update(&context, buffer, i);
   }
   digest=SHA1Final(&context);
   fclose(file);
   for (i = 0; i < 5; i++) {
   for (j = 0; j < 4; j++) {
   printf("%02X", digest[i*4+j]);
   }
   putchar(' ');
   }
   putchar('\n');
   exit(0);
   }

 */
